### PR TITLE
Add test for `s3/writer` and Refactor.

### PR DESCRIPTION
**Pull request by hlts2**
Signed-off-by: hlts2 <hiroto.funakoshi.hiroto@gmail.com>

### Description:

I refactored and added the tests for `s3/writer`.
And when I rebase PR, I found a build error. So I fixed it.

#### Refactor

- WHY
I can not write the tests for this method because the object is created internally and really need to connect to s3.
https://github.com/vdaas/vald/blob/master/internal/db/storage/blob/s3/writer/writer.go#L99-L122 

- WHAT
so I want to rewrite to be able to mock.
For example. I’m thinking of making an s3manager wrapper like #659  and #622 .

#### Related issue

- refactor: https://github.com/vdaas/vald/projects/3#card-44992105
- test: https://github.com/vdaas/vald/projects/3#card-43770144

***Autor Review Required***

#### Review Step

- [x] Refactor 
- [x] SET Team Review for refactoring contents
- [x] Author Review for refactoring contents
- [x] Add tests
- [ ] Review test code 

### Environment:

- Go Version: 1.15.1
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.1

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [x] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [x] I have added tests and benchmarks to cover my changes.
- [x] I have ensured all new and existing tests passed.
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: internal/db/storage/blob/s3/sdk/s3/s3manager/s3manager.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	// Uploader is type alias of s3manager.Uploader.
+	Uploader = s3manager.Uploader
+	// UploadInput is type alias of s3manager.UploadInput.
+	UploadInput = s3manager.UploadInput
+	// UploadOutput is type alias of s3manager.UploadOutput.
+	UploadOutput = s3manager.UploadOutput
+)
+
+// Upload represents an interface to upload to s3.
+type Upload interface {
```
Comment (kpango) [inline internal/db/storage/blob/s3/sdk/s3/s3manager/s3manager.go:19]: how about this
```suggestion
type UploadClient interface {
```

Comment (hlts2) [reply internal/db/storage/blob/s3/sdk/s3/s3manager/s3manager.go:19]: It seems good!
thank you.

**Review by kevindiu — DISMISSED** [review]
LGTM

File: internal/db/storage/blob/s3/writer/writer_test.go [hunk 2]
```diff
@@ -51,36 +61,44 @@ func TestNew(t *testing.T) {
...
+					eg:          errgroup.Get(),
+					contentType: "application/octet-stream",
+					maxPartSize: 64 * 1024 * 1024,
+					s3manager:   s3manager.New(),
+				},
+			},
+		},
+
+		{
+			name: "returns writer and outputs the warn when option is not empty",
```
Comment (kevindiu) [inline internal/db/storage/blob/s3/writer/writer_test.go:80]: 🤔 it is difficult to check if the warning message is outputted or not 🤣

File: internal/db/storage/blob/s3/writer/writer_test.go [hunk 3]
```diff
@@ -132,54 +150,34 @@ func Test_writer_Open(t *testing.T) {
...
-		           pw: nil,
-		           wg: sync.WaitGroup{},
-		           },
-		           want: want{},
-		           checkFunc: defaultCheckFunc,
-		       }
-		   }(),
-		*/
+		{
+			name: "returns nil when no error occurs",
```
Comment (kevindiu) [inline internal/db/storage/blob/s3/writer/writer_test.go:154]: can you add a test case for error case?

Comment (kevindiu) [reply internal/db/storage/blob/s3/writer/writer_test.go:154]: btw do you know why we use waitgroup here?
https://github.com/vdaas/vald/pull/672/files#diff-da7d1c78589548bbb90ba2560947237bR69

File: internal/db/storage/blob/s3/writer/writer_test.go (hunk 2 above)
Comment (hlts2) [reply internal/db/storage/blob/s3/writer/writer_test.go:80]: I think so too.
If the writer returns an error, I will fix it

File: internal/db/storage/blob/s3/writer/writer_test.go (hunk 3 above)
Comment (kevindiu) [reply internal/db/storage/blob/s3/writer/writer_test.go:154]: please ignore this comment 
`btw do you know why we use waitgroup here?`

**Comment by kevindiu** [comment]
Can you update the test case to cover this logic?
https://github.com/vdaas/vald/blob/b075b0f6520add630a1ffc2ea5a4479847519609/internal/db/storage/blob/s3/writer/writer.go#L93

**Review by kevindiu — APPROVED** [review]
LGTM

File: internal/db/storage/blob/s3/sdk/s3/s3manager/s3manager.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+type s3mngr struct{}
+
+// New returns S3Manager implementation.
+func New() S3Manager {
+	return new(s3mngr)
+}
+
+func (*s3mngr) NewUploaderWithClient(svc s3iface.S3API, options ...func(*Uploader)) UploadClient {
```
Comment (vankichi) [inline internal/db/storage/blob/s3/sdk/s3/s3manager/s3manager.go:35]: - could you please write comment for `NewUploadWithClient()` ?

**Review by kevindiu — APPROVED** [review]
LGTM

File: internal/db/storage/blob/s3/writer/writer_test.go [hunk 5]
```diff
@@ -132,54 +150,34 @@ func Test_writer_Open(t *testing.T) {
...
+								}, nil
+							},
+						}
+					},
+				},
+			},
+			want: want{
+				err: nil,
+			},
+		},
```
[the remark this answers is not in the document]
Comment (hlts2) [reply internal/db/storage/blob/s3/writer/writer_test.go:175]: sorry. There are no errors in this code. So I haven't listed it.

**Review by kevindiu — APPROVED** [review]
LGTM

**Review by vankichi — APPROVED** [review]

**Pull request merged**